### PR TITLE
DDF-4552 Migrate search-settings to React

### PR DESCRIPTION
**Pull request by Corey-Collins**
#### What does this PR do?
Migrates the search-settings component from marionette to react
#### Who is reviewing it? 
@willwill96 
@andrewzimmer 
@samuelechu 
@leo-sakh 
@hayleynorton 

#### Select relevant component teams: 
@codice/ui 
#### Ask 2 committers to review/merge the PR and tag them here.

@djblue 
@andrewkfiedler 

#### How should this be tested?

Test search-settings in dropdown and in user-settings is the same behavior as the marionette component

**In dropdown area:**
 - found in search > (click three dots) > settings
 - set a new result count
 - verify save | cancel buttons are in footer
 - cancel without saving
 - go back to dropdown, verify old value is there
 - change value again
 - save changes, verify confirmation
 - verify changes reflect search
 - verify new values are in user settings > search settings

**In user settings:**

 - found by click gear icon in top right > search settings
 - verify save | cancel buttons are not in footer
 - change value and click back button
 - verify changes are saved after you clicked the back button
 - verify changes are reflected in search settings dropdown area
#### What are the relevant tickets?
For GH Issues:
Fixes: #4552 

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+type Props = {
+  onClose?: () => void
+  model?: any
+  showFooter: boolean
+} & WithBackboneProps
+
+export default hot(module)(
+  withListenTo(
+    class SearchSettings extends React.Component<Props, {}> {
+      propertyModel: any
```
Comment (leo-sakh) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: ❓ Do you really need to expose these properties as a class member variables? It appears that in most/all cases you're just using the references to get to values that can just be memoized in a function instead.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      queryModel: any
+      querySettingsView: any
+      propertyView: any
+      constructor(props: Props) {
```
Comment (leo-sakh) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: ✏️ I don't think you need a constructor here (or even to extend a `React.Component` for that matter. You're not using state so you can just pull up most of the functions out and make this a stateless component.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+          .get('querySettings')
+          .set(this.querySettingsView.toJSON())
+      }
+      triggerCancel = () => {
+        this.setPropertyView()
+        this.setQuerySettingsView()
+        if (!!this.props.onClose) {
+          this.props.onClose()
+        }
+        this.forceUpdate()
```
Comment (leo-sakh) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: ❓ Why is this necessary here?

Comment (willwill96) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: I think it's used to re-render the QuerySettingsView. We should be keeping that view in the state since it directly affects rendering

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  padding: ${props => props.theme.minimumSpacing}
+    ${props => props.theme.minimumSpacing};
+`
+
+const PropertyResultCount = styled.div`
+  margin-bottom: ${props => props.theme.largeSpacing};
+`
+
+const QuerySettings = styled.div`
+  .editor-header,
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: ✏️ We should remove class selectors and use styled components instead

Comment (Corey-Collins) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: .editor-header .editor-footer and editor-properties are within the marionette container. not sure what else to do in that case but individual styled components doesn't seem possible

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx (hunk 3 above)
Comment (Corey-Collins) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: i tried putting the views in the state, and having a setState that re-instantiates those view objects but it still doesn't reset the value.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx (hunk 4 above)
Comment (adimka) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: This is an interesting one - we don't really want to migrate these styles onto property view as they're only used here. This will be easy once `property.view` is converted as we can just use a `styled(Property)` extension.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx (hunk 2 above)
Comment (Corey-Collins) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: it seems we will need to extend `React.Component` since this class uses the `ComponentWillMount` lifecycle function

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+type Props = {
+  onClose?: () => void
+  model?: any
+  showFooter: boolean
+} & WithBackboneProps
+
+export default hot(module)(
+  withListenTo(
+    class SearchSettings extends React.Component<Props, {}> {
```
Comment (djblue) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/search-settings/search-settings.tsx]: Can we pull out the class into the module scope to reduce the nesting. We could do:

`export default hot(module)(withListenTo(SearchSettings))` 

instead.

**Comment by bdeining** [comment]
build now

**Comment by samuelechu** [comment]
Hero successful: verified existing behavior is preserved and that changes are reflected in search.

Some suggestions that are out of scope for this ticket: 
1. Rerender the search form when the search settings are changed, rather than having to switch between search forms or perform a search for the change in settings to appear. This might be a bug, it seems like you could just listen to changes on the user model to do this.
2.  Dropdown search settings don't currently sync with sidebar search settings changes - also seems like this could be fixed by listening to the correct model.
3. Show `Cancel` button only when changes have been made.
